common/src/main/java/com/microsoft/identity/common/internal/platform/DevicePopManager.java

```diff
@@ -0,0 +1,834 @@
+
+        /**
+         * The HTTP method.
+         */
+        private static final String HTTP_METHOD = "m";
+
+        /**
+         * The host part of the resource server URL.
+         */
+        private static final String HTTP_HOST = "u";
+
+        /**
+         * The path part of the resource server URL.
+         */
+        private static final String HTTP_PATH = "p";
+
+        /**
+         * The JWK signed into this JWT.
+         */
+        private static final String CNF = "cnf";
+
+        /**
+         * A random value used for replay protection.
+         */
+        private static final String NONCE = "nonce";
+    }
+
+    /**
+     * Algorithms supported by this KeyPairGenerator.
+     */
+    static final class KeyPairGeneratorAlgorithms {
+        static final String RSA = "RSA";
+    }
+
+    DevicePopManager() throws KeyStoreException, CertificateException,
+            NoSuchAlgorithmException, IOException {
+        mKeyStore = KeyStore.getInstance(ANDROID_KEYSTORE);
+        mKeyStore.load(null);
+    }
+
+    @Override
+    public boolean asymmetricKeyExists() {
+        boolean exists = false;
+
+        try {
+            exists = mKeyStore.containsAlias(KEYSTORE_ENTRY_ALIAS);
+        } catch (final KeyStoreException e) {
+            Logger.error(
+                    TAG,
+                    "Error while querying KeyStore",
+                    e
+            );
+        }
+
+        return exists;
+    }
+
+    @Override
+    public boolean asymmetricKeyExists(@NonNull final String thumbprint) {
+        if (asymmetricKeyExists()) { // Test if keys exist at all...
+            try {
+                return getAsymmetricKeyThumbprint().equals(thumbprint);
+            } catch (final ClientException e) {
+                Logger.error(
+                        TAG,
+                        "Error while comparing thumbprints.",
+                        e
+                );
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public String getAsymmetricKeyThumbprint() throws ClientException {
+        final Exception exception;
+        final String errCode;
+
+        try {
+            final KeyStore.Entry keyEntry = mKeyStore.getEntry(KEYSTORE_ENTRY_ALIAS, null);
+            final KeyPair rsaKeyPair = getKeyPairForEntry(keyEntry);
+            final RSAKey rsaKey = getRsaKeyForKeyPair(rsaKeyPair);
+            return getThumbprintForRsaKey(rsaKey);
+        } catch (final KeyStoreException e) {
+            exception = e;
+            errCode = KEYSTORE_NOT_INITIALIZED;
+        } catch (final NoSuchAlgorithmException e) {
+            exception = e;
+            errCode = NO_SUCH_ALGORITHM;
+        } catch (final UnrecoverableEntryException e) {
+            exception = e;
+            errCode = INVALID_PROTECTION_PARAMS;
+        } catch (final JOSEException e) {
+            exception = e;
+            errCode = THUMBPRINT_COMPUTATION_FAILURE;
+        }
```

Comment:
should we also catch other unexpected error?

Reply:
bringing this up because it looks like we want to only throw ClientException here if anything bad happens. I may be wrong.

Reply:
actually, nvm.ExceptionAdapter.baseExceptionFromException() should do the trick.